src/utilities/attachments-manager.ts

```diff
@@ -0,0 +1,89 @@
+  }
+
+  public isAnonymous() {
+    // The client will be anonymous if firebaseJwt undefined
+    return !this.firebaseJwt;
+  }
+
+  public getSessionId() {
+    return this.sessionId;
+  }
+
+  public async createFolder(interactiveId: string): Promise<IWritableAttachmentsFolder> {
+    const folderResource = await this.tokenServiceClient.createResource({
+      tool: kTokenServiceToolName,
+      type: "s3Folder",
+      name: `${this.learnerId}-${interactiveId}`,
+      description: "attachment",
+      accessRuleType: this.isAnonymous() ? "readWriteToken" : "user"
+    });
+    this.resources[folderResource.id] = folderResource as S3Resource;
+    return {
+      id: folderResource.id,
+      readWriteToken: this.tokenServiceClient.getReadWriteToken(folderResource) || undefined
+    };
+  }
+
+  public async getSignedWriteUrl(
+    folder: IAttachmentsFolder, name: string, expires = kDefaultWriteExpirationSec
+  ): Promise<[string, IReadableAttachmentInfo]> {
+    const folderResource = await this.getFolderResource(folder);
+    const publicPath = this.tokenServiceClient.getPublicS3Path(folderResource, `${this.sessionId}/${name}`);
+    const url = await this.getSignedUrlFromPublicPath(folder, publicPath, "putObject", expires);
+    // returns the writable url and the information required to read it
+    return [url, { folder: { id: folder.id }, publicPath }];
+  }
+
+  public getSignedReadUrl(attachmentInfo: IReadableAttachmentInfo, expires = kDefaultReadExpirationSec) {
+    const { folder, publicPath } = attachmentInfo;
+    return this.getSignedUrlFromPublicPath(folder, publicPath, "getObject", expires);
+  }
+
+  private async getFolderResource(folder: IAttachmentsFolder): Promise<S3Resource> {
+    let folderResource: S3Resource = this.resources[folder.id];
+    if (!folderResource) {
+      folderResource = await this.tokenServiceClient.getResource(folder.id) as S3Resource;
+      this.resources[folderResource.id] = folderResource;
+    }
+    return folderResource;
+  }
+
+  private getCredentials(folder: IAttachmentsFolder): Promise<Credentials> {
+    const readWriteToken = isWritableAttachmentsFolder(folder) ? folder.readWriteToken : undefined;
+    return this.tokenServiceClient.getCredentials(folder.id, readWriteToken);
+  }
+
+  private async getSignedUrlFromPublicPath(folder: IAttachmentsFolder, publicPath: string, operation: string, expires: number) {
+    const folderResource = await this.getFolderResource(folder);
+    const credentials = await this.getCredentials(folder);
+    const { bucket, region } = folderResource;
+    const { accessKeyId, secretAccessKey, sessionToken } = credentials;
+    const s3 = new AWS.S3({ region, accessKeyId, secretAccessKey, sessionToken });
+    // https://zaccharles.medium.com/s3-uploads-proxies-vs-presigned-urls-vs-presigned-posts-9661e2b37932
+    const s3UrlParams = { Bucket: bucket, Key: publicPath, Expires: expires };
```

Comment:
My understanding is that the `ContentType` param must be set here so it matches the content type sent in the PUT request. When the presigned url fails AWS shows the content that should be signed and it includes this content type.  For cfm example when it sends data it is: `text/plain;charset=UTF-8`.

To make that possible I think it means the interactive api needs to be changed to require sending this content type when it requests the signed url. But before making that change I'd recommend testing this out locally by hard coding it.

Evidence that it is required:
https://stackoverflow.com/a/32464938
https://github.com/aws/aws-sdk-js/blob/307e82673b48577fce4389e4ce03f95064e8fe0d/test/browser/sample/s3upload.html#L47-L55

Entry point in the source (might be slightly different version of library)
https://github.com/aws/aws-sdk-js/blob/7872257f42daafeba691fccd769922100fea05be/lib/services/s3.js#L924
I tried tracing this through by just looking at the source but got lost and could only find the requests that work with the GET operations.

Reply:
I did also see a note that I lost now how when the POST operation is used so the content type is not required.

Reply:
Here is where the string to sign is computed:
https://github.com/aws/aws-sdk-js/blob/e9371bb4ea0671b8034cef08fcdd69bf434d7a03/lib/signers/s3.js#L66-L86

So it always has a content type or a blank line. And it always has an MD5 hash or a blank line. In the error message we get back from AWS we can see the blank line of the MD5 hash followed by the content type, then the expiration time. So that all gives me more confidence that what we are missing is the ContentType in the presigned url params. 

```
PUT

text/plain;charset=UTF-8
1627651148
...
```

I also found this blog post that makes this pretty explicit:
https://advancedweb.hu/differences-between-put-and-post-s3-signed-urls/#:~:text=for%20put%20urls%20the%20signing

Also the conclusion of that blog post is that POST urls are better. Because they let you use wildcard content types, and you can set a size restriction on the upload.  However if I remember right @kswenson you found a blog post suggesting PUT is better.

Reply:
>So that all gives me more confidence that what we are missing is the ContentType in the presigned url params.

Good sleuthing -- I'll give that a try.

>Also the conclusion of that blog post is that POST urls are better. Because they let you use wildcard content types, and you can set a size restriction on the upload. However if I remember right @kswenson you found a blog post suggesting PUT is better.

I'll take a look at that article; I hadn't seen it before. Actually, the article I had found also suggested that POST URLs are better for the same reasons, but then the examples given were for uploading a File using POST or arbitrary content using PUT and it wasn't immediately obvious how to upload arbitrary content using POST so I went with the more salient example.